front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts

```diff
@@ -285,15 +288,35 @@ const importTimetable = async (
 
   nodes = convertGeoCoords(nodes);
 
+  const DTOLabels: NGELabel[] = [];
   // Create one NGE train run per OSRD train schedule
-  const trainruns: Trainrun[] = trainSchedules.map((trainSchedule) => ({
-    id: trainSchedule.id,
-    name: trainSchedule.train_name,
-    categoryId: DEFAULT_TRAINRUN_CATEGORY.id,
-    frequencyId: DEFAULT_TRAINRUN_FREQUENCY.id,
-    trainrunTimeCategoryId: DEFAULT_TRAINRUN_TIME_CATEGORY.id,
-    labelIds: [],
-  }));
+  const trainruns: Trainrun[] = trainSchedules.map((trainSchedule) => {
+    let formatedLabels: NGELabel[] = [];
+    if (trainSchedule.labels) {
+      formatedLabels = trainSchedule.labels.map((label) => {
+        const DTOLabel = DTOLabels.find((DTOlabel) => DTOlabel.label === label);
+        if (DTOLabel) {
+          return DTOLabel;
+        }
+        const newDTOLabel: NGELabel = {
+          id: uuid(),
+          label,
+          labelGroupId: 1,
+          labelRef: 'Trainrun',
+        };
+        DTOLabels.push(newDTOLabel);
+        return newDTOLabel;
+      });
+    }
+    return {
+      id: trainSchedule.id,
+      name: trainSchedule.train_name,
+      categoryId: DEFAULT_TRAINRUN_CATEGORY.id,
+      frequencyId: DEFAULT_TRAINRUN_FREQUENCY.id,
+      trainrunTimeCategoryId: DEFAULT_TRAINRUN_TIME_CATEGORY.id,
+      labelIds: formatedLabels.map((label) => label.id),
+    };
+  });
```

Comment:
To avoid doing another map at the end, maybe we can directly stock the ids in `formatedLabels` ?
I also feel like we have a lot of nesting loops. Maybe we can improve that ?

```suggestion
  const trainruns: Trainrun[] = trainSchedules.map((trainSchedule) => {
    let formatedLabels: (string | number)[] = [];
    if (trainSchedule.labels) {
      formatedLabels = trainSchedule.labels.map((label) => {
        const DTOLabel = DTOLabels.find((DTOlabel) => DTOlabel.label === label);
        if (DTOLabel) {
          return DTOLabel.id;
        }
        const newDTOLabel: NGELabel = {
          id: uuid(),
          label,
          labelGroupId: 1,
          labelRef: 'Trainrun',
        };
        DTOLabels.push(newDTOLabel);
        return newDTOLabel.id;
      });
    }
    return {
      id: trainSchedule.id,
      name: trainSchedule.train_name,
      categoryId: DEFAULT_TRAINRUN_CATEGORY.id,
      frequencyId: DEFAULT_TRAINRUN_FREQUENCY.id,
      trainrunTimeCategoryId: DEFAULT_TRAINRUN_TIME_CATEGORY.id,
      labelIds: formatedLabels,
    };
  });
```

Reply:
Done